### PR TITLE
Feature/boas 1137 unpublished changes

### PR DESCRIPTION
**Pull request by HadrianHughesKC**
## Describe your changes

* feat(api/applications): add a column for cases with unpublished changes
* feat(api/applications): implement updating hasUnpublishedChanges field
Check if any fields which were updated are fields which will eventually be
published. If so, set `hasUnpublishedChanges` to true.
* feat(api/applications): set hasUnpublishedChanges to false on publish
* feat(api/applications): return hasUnpublishedChanges in response
* feat(web/applications): render banner for unpublished changes
* chore(web/applications): optimistically set hasUnpublishedChanges to false
* test(api/applications): update test for publishing application
* chore(api/applications): compare stringified case events to diff changes

Tested locally.

## Issue ticket number and link

[BOAS-1137](https://pins-ds.atlassian.net/browse/BOAS-1137)

## Type of change 🧩

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Other (please explain in the description section above)

## Checklist before requesting a review

- [x] I have performed a self-review of my own code
- [x] I have double checked this work does not include any hardcoded secrets or passwords
- [ ] I have made corresponding changes to the documentation
- [x] I have provided details on how I have tested my code
- [x] I have referenced the ticket number above
- [ ] I have provided a description of the ticket
- [ ] I have included unit tests to cover any testable code changes

[BOAS-1137]: https://pins-ds.atlassian.net/browse/BOAS-1137?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/api/src/server/applications/application/application.controller.js [hunk 1]
```diff
@@ -49,17 +50,49 @@ export const createApplication = async (request, response) => {
...
 		applicantId: head(body?.applicants)?.id,
 		...mappedApplicationDetails
 	});
 
 	if (!updateResponse) {
-		throw new Error('Application not found');
+		throw new BackOfficeAppError('Application not found', 500);
+	}
+
+	const publishableFieldsChanged = publishedFieldsHaveChanged(originalResponse, updateResponse);
```
Comment (beejjacobs) [inline apps/api/src/server/applications/application/application.controller.js:79]: Is it worth checking `hasUnpublishedChanges`? If it's already true we don't need to check again.

File: apps/api/src/server/applications/key-dates/key-dates.controller.js [hunk 2]
```diff
@@ -38,9 +39,25 @@ export const updateKeyDates = async ({ body, params }, response) => {
 
 	const keyDates = mapRequestToKeyDates(body);
 
+	const originalCase = await caseRepository.getById(id, {
```
Comment (beejjacobs) [inline apps/api/src/server/applications/key-dates/key-dates.controller.js:42]: This looks similar to `application.controller.js` - are there some service or repo methods we should be adding to hold this logic?

File: apps/api/src/server/utils/published-fields-changed.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+/** @typedef {import('@pins/applications.api').Schema.Case} Case */
+
+/**
+ * Checks whether the given column name is included when the case is published
+ *
+ * @param {Case} original
+ * @param {Case} updated
+ * @returns {boolean}
+ * */
+export const publishedFieldsHaveChanged = (original, updated) => {
```
Comment (beejjacobs) [inline apps/api/src/server/utils/published-fields-changed.js:12]: This looks like it'd work fine, but if we change the schema/payload we have to update this too.

What if instead we build an event with the before and after cases, and compare them (i.e. using `buildNsipProjectPayload`)? Could even just stringify and check if they're equal. That way it's tied into what we publish.

**Review by beejjacobs — CHANGES_REQUESTED** [review]
Looks good, just got some suggestions/ideas.

Also, should there be a swagger/api-spec update for this?

File: apps/api/src/server/utils/published-fields-changed.js (hunk 3 above)
Comment (HadrianHughesKC) [reply apps/api/src/server/utils/published-fields-changed.js]: Yeah good idea, I've made that change. We just need to make sure the `buildNsipProjectPayload` function only includes properties that should be in the event. There's some `@ts-ignore` dotted around that might allow extra properties to be included. I think it's right for now though.

File: apps/api/src/server/applications/key-dates/key-dates.controller.js (hunk 2 above)
Comment (HadrianHughesKC) [reply apps/api/src/server/applications/key-dates/key-dates.controller.js]: I've made the change below and done it in a common util so there's no duplication.

File: apps/api/src/server/utils/published-fields-changed.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+/** @typedef {import('@pins/applications.api').Schema.Case} Case */
+
+/**
+ * Checks whether the given column name is included when the case is published
+ *
+ * @param {Case} original
+ * @param {Case} updated
+ * @returns {boolean}
+ * */
+const publishedFieldsHaveChanged = (original, updated) => {
```
Comment (beejjacobs) [inline apps/api/src/server/utils/published-fields-changed.js:14]: I think its worth making it clear in the name this is for cases, not generic.

`publishedCaseFieldsHaveChanged` perhaps?

**Review by beejjacobs — COMMENTED** [review]
Looks good. One minor comment below. Also does this change the API returned value? (i.e. do we include `hasUnpublishedChanges`? If so I'd expected a swagger-output change.

File: apps/api/src/server/utils/published-fields-changed.js (hunk 4 above)
Comment (HadrianHughesKC) [reply apps/api/src/server/utils/published-fields-changed.js]: That's done

**Comment by HadrianHughesKC** [comment]
> Looks good. One minor comment below. Also does this change the API returned value? (i.e. do we include `hasUnpublishedChanges`?) If so I'd expected a swagger-output change.

Thanks @beejjacobs I've regenerated swagger.

**Pull request merged**